internal/evaluator/rego.go

```diff
@@ -0,0 +1,120 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+// IMPORTANT: The rego functions in this file never return an error. Instead, they return no value
+// when an error is encountered. If they did return an error, opa would exit abruptly and it would
+// not produce a report of which policy rules succeeded/failed.
+
+package evaluator
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+	"io"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/open-policy-agent/opa/types"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/enterprise-contract/ec-cli/internal/fetchers/oci"
+)
+
+const ociBlobName = "ec.oci.blob"
+
+func registerOCIBlob() {
+	decl := rego.Function{
+		Name: ociBlobName,
+		Decl: types.NewFunction(
+			types.Args(
+				types.Named("ref", types.S).Description("OCI blob reference"),
+			),
+			types.Named("blob", types.S).Description("the OCI blob"),
+		),
+		Memoize:          true,
+		Nondeterministic: true,
+	}
+
+	rego.RegisterBuiltin1(&decl, ociBlob)
+}
+
+const maxBytes = 10 * 1024 * 1024 // 10 MB
+
+func ociBlob(bctx rego.BuiltinContext, a *ast.Term) (*ast.Term, error) {
+	uri, ok := a.Value.(ast.String)
+	if !ok {
+		return nil, nil
+	}
+
+	ref, err := name.NewDigest(string(uri))
+	if err != nil {
+		log.Errorf("%s new digest: %s", ociBlobName, err)
+		return nil, nil
+	}
+
+	opts := []remote.Option{
+		remote.WithTransport(remote.DefaultTransport),
+		remote.WithContext(bctx.Context),
+		remote.WithAuthFromKeychain(authn.DefaultKeychain),
+	}
+
+	rawLayer, err := oci.NewClient(bctx.Context).Layer(ref, opts...)
+	if err != nil {
+		log.Errorf("%s fetch layer: %s", ociBlobName, err)
+		return nil, nil
+	}
+
+	layer, err := rawLayer.Uncompressed()
+	if err != nil {
+		log.Errorf("%s layer uncompressed: %s", ociBlobName, err)
+		return nil, nil
+	}
+	defer layer.Close()
+
+	// TODO: Other algorithms are technically supported, e.g. sha512. However, support for those is
+	// not complete in the go-containerregistry library, e.g. name.NewDigest throws an error if
+	// sha256 is not used. This is good for now, but may need revisiting later.
+	hasher := sha256.New()
+	// Setup some safeguards. First, use LimitReader to avoid an unbounded amount of data from being
+	// read. Second, use TeeReader so we can compute the digest of the content read.
+	reader := io.TeeReader(io.LimitReader(layer, maxBytes), hasher)
+
+	var blob bytes.Buffer
+	if _, err := io.Copy(&blob, reader); err != nil {
+		log.Errorf("%s copy buffer: %s", ociBlobName, err)
+		return nil, nil
+	}
+
+	sum := fmt.Sprintf("sha256:%x", hasher.Sum(nil))
+	// io.LimitReader truncates the layer if it exceeds its limits. This ensures unexpected behavior
```

Comment:
I might be misunderstanding it, but...

```suggestion
	// io.LimitReader truncates the layer if it exceeds its limits. This avoids unexpected behavior
```

Reply:
I re-wrote this sentence because it was just awful.